lib/fsm/utils.go

```diff
@@ -41,9 +41,44 @@ func CanRollback(plan *storage.OperationPlan, phaseID string) error {
 		return trace.BadParameter(
 			"phase %q has already been rolled back", phase.ID)
 	}
+	if !latestRollback(plan.Phases, phaseID) {
+		return trace.BadParameter(
+			"rollback subsequent phases before rolling back phase %q", phase.ID)
+	}
 	return nil
 }
 
+// latestRollback returns true if the phase specified by phaseID is next in line
+// to be rolled back.
+func latestRollback(phases []storage.OperationPhase, phaseID string) bool {
+	if len(phases) == 0 {
+		return false
+	}
+
+	// latest keeps track of the latest phase that has not yet been rolled back.
+	var latest storage.OperationPhase
+
+L:
+	for _, phase := range phases {
+		if latestRollback(phase.Phases, phaseID) {
```

Comment:
Why do we need recursion here? I would make the algorithm very simple, all we need to do is to "unwrap" the plan (we have the method called "FlattenPlan" somewhere, there's also a method called GetLeafPhases on the plan itself IIRC) and then check if any of the phases that appear after phaseID have been rolled back or are unstarted. Sounds simple enough and easy to reason about, what do you think?

Reply:
And I think this is good. If I'm not mistaken, rolling back top-level phases is not implemented properly at the moment, as in it starts rollback top-down, not in reverse order. Or it may be disabled completely, I don't recall.

Reply:
Oh I see. Okay, we can update to only allow rollback of leaf phases.